scipy/_lib/_testutils.py

```diff
@@ -326,41 +328,104 @@ def load(modname):
+def run_in_parallel(
+        wrapped=None, n_workers=10, use_barrier=True, assert_all_equal=True,
+        pass_thread_id=False):
     """
-    Run a given function concurrently across a given number of threads.
+    Decorates a given function to execute it concurrently
+    across a given number of threads.
 
-    This is equivalent to using a ThreadPoolExecutor, but using the threading
-    primitives instead. This function ensures that the closure passed by
-    parameter gets called concurrently by setting up a barrier before it gets
-    called before any of the threads.
+    Usually, this decorator is used to mark tests for concurrent execution of
+    functions or classes against a fixed set of inputs, whose results are
+    compared as well:
+
+    .. code-block:: python
+
+        import pytest
+        import numpy as np
+        from scipy._lib._testutils import run_in_parallel
+
+        class ExampleClass:
+            def __call__(self, x):
+                return x + 5
+
+        @pytest.fixture
+        def fixed_inputs():
+            example_input = np.ones(4, 5)
+            return ExampleClass(), example_input
+
+        @run_in_parallel
+        def test_concurrent(fixed_inputs):
+            # Check that ExampleClass does not mutate under concurrency
+            # and that concurrent calls yield the same return value
+            to_call, input = fixed_inputs
+            return to_call(input)
+
+    This decorator is interoperable with pytest fixtures and markers as well:
+
+    .. code-block:: python
+
+        @pytest.mark.parametrize('param', [0, 1])
+        @run_in_parallel
+        def test_concurrent(fixed_inputs, param):
+            to_call, input = fixed_inputs
+            return to_call(input)
+
+    Also, the thread number can be obtained by setting `pass_thread_id=True`
+    and defining the `thread_id` kwarg in the target function:
+
+    .. code-block:: python
+
+        @run_in_parallel(pass_thread_id=True)
+        def test_concurrent(thread_id=0):
+            # thread_id has the thread number that is executing this function.
+            pass
+
+    Finally, it is possible to opt-out from output comparison by setting
+    `assert_all_close=False`
 
     Arguments
     ---------
     n_workers: int
-        Number of concurrent threads to spawn.
-    fn: callable
-        Function closure to execute concurrently. Its first argument will
-        be the thread id.
-    *args: tuple
-        Variable number of positional arguments to pass to the function.
-    **kwargs: dict
-        Keyword arguments to pass to the function.
+        Number of concurrent threads to spawn. Default: 10
+    use_barrier: bool
+        If `True`, then a barrier is set before all threads start executing the
+        target function. Default: `True`
+    assert_all_equal: bool
+        If `True`, then the return value produced by the function will be
+        compared for equality across all executing threads. Default: `True`
```

Comment:
Minor: the name hints that it checks equality but the actual test is `assert_allclose` with the default tolerance. Before addressing this small inconsistency, could you help me understand the intended usages, is it:

- run a bunch of identical workloads concurrently, check that that all results are identical
- run a bunch of possibly different workloads, do not check anything

Is this enough?

Reply:
Both should be acceptable workloads, for instance, consider the test defined in `test_rgi.py` 

https://github.com/andfoy/scipy/blob/abcf2e4981070a70e9d32ad0ad927ca75ff08a06/scipy/interpolate/tests/test_rgi.py#L703

This test checks for a not mutating attribute of a class that gets called from multiple threads with different arguments, in this case, checking for equality does not make sense